### DISCUSSION TITLE
Using .set_dependencies | returning values from tasks |  using returned values in 2 or more subsequent tasks

### DISCUSSION THREAD

**Discussion by ReubenTheDS** [PrefectHQ/prefect#3410] [Q&A] [answered]
Is it possible to have tasks run sequentially like so, while also storing their return values

```
from prefect import task, Flow


@task(log_stdout=True)
def task_1():
    print("In task1")
    #create some files
    return 1



@task(log_stdout=True)
def task_2(val_1):
    print("In task2")
    print("val1: {}".format(val_1))
    return 2



@task(log_stdout=True)
def task_3(val_1, val_2):
    print("In task3")
    print("val1: {}, val_2:{}".format(val_1, val_2))
    return 3



# define Prefect's flow
with Flow("Seq_flow_with_args") as flow:

    # I need to store the result of task_1 here
    # to send to task_2 & task_3 later


    # I need to store the result of task2 here
    flow.set_dependencies(
        task=task_2(result_of_task1),
        upstream_tasks=[task_1])  # since task_2 will read the files generated by task_1


    # I need to send task3 2 results: result_of_task1, result_of_task1
    flow.set_dependencies(
        task=task_3(result_of_task1, result_of_task2),
        upstream_tasks=[task_2])

flow.run()
```

**Comment by joshmeek** [comment, marked as answer]
Hi @ReubenTheDS yes this is supported! Take a look at the documentation on [persistence](https://docs.prefect.io/core/concepts/persistence.html) and [results](https://docs.prefect.io/core/concepts/results.html)! This will allow you to store the results from your tasks so that they can be accessed in later runs or even restarts of the same run.

Not sure if I full follow your question about setting dependencies but you can use the results from tasks in other tasks using something like this:

```python
with Flow("Seq_flow_with_args") as flow:
    t1 = task_1()
    t2 = task_2(t1)
    t3 = task_3(t1, t2)
```

The documentation also contains examples of using the imperative API along with the functional API if that is what you are after [[1]](https://docs.prefect.io/core/getting_started/first-steps.html#imperative-api) [[2]](https://docs.prefect.io/core/idioms/task-results.html)

**Reply by ReubenTheDS** [reply]
Thanks for the answer @joshmeek .
I already know that you can do this:
```
t1 = task_1()
t2 = task_2( t1 )
```

What I need is something like this:
```
t1 = task_1() # task 1 generates files( this will take time ) as well as returns something
t2 = task_2( t1 ) # and I want task_2 to run only after task_1 is completed, since it will read the files generated
t3 = task_3( t1, t2 ) # and I want task_3 to run only after task_2 is completed, for reason X
...
```

I've looked at the refs mentioned, but I wasn't able to find any code in there, or in the repository, where a task should run after an"other", as well as get an input from that "other" task.

The above is obviously pseudo-code and I wanted to know whether you can have many tasks executing sequentially, as well as use each other's return values. 

I'm guessing that's not the way Prefect should run ideally, but that's the ask :) 

PS: If you're going to point me to edges/triggers, please let me know if the above can be done, again. Phew.

**Reply by joshmeek** [reply]
@ReubenTheDS what you described above in your code snippet is exactly how Prefect flows operate. Task 2 won't run until task 1 completes and task 3 won't run until task 1 and task 2 are complete.

**Reply by ReubenTheDS** [reply]
Thanks @joshmeek 
OK, I should have been clearer :)

Is it possible to have this:
```
with Flow("Seq_flow_with_args") as flow:
    t1 = task_1() # task 1 generates files( this will take time ) as well as returns something
    t2 = task_2( t1 ) # and I want task_2 to run only after task_1 is completed, since it will read the files generated
    t3 = task_3( t1, t2 ) # and I want task_3 to run only after task_2 is completed, for reason X
    t4 = task_4()  # wait for all above tasks to finish, without passing t1, t2 or t3
```

This is where I originally wanted more clarity on ```.set_dependencies()``` on either ```flow``` or the tasks themselves

**Reply by joshmeek** [reply]
Since you are passing data from task to task all of what you described for t1, t2, and t3 looks good. Now for t4 you would want to call something like:

```python
with Flow().....
    ...    
    t4.set_upstream(t1)
    t4.set_upstream(t2)
    t4.set_upstream(t3)
```

Now task 4 won't start until all of the upstream tasks finish. As an alternative you could also do something like this:

```python
with Flow().....
    ...    
    t1.set_downstream(t4)
    t2.set_downstream(t4)
    t3.set_downstream(t4)
```

**Reply by ReubenTheDS** [reply]
Thank you @joshmeek This works.
